hal/src/common/gpio/v2.rs

```diff
@@ -0,0 +1,1701 @@
+            impl<I> From<PinT<I, $Mode1>> for PinT<I, $Mode2>
+            where
+                I: PinId,
+            {
+                #[inline]
+                fn from(pin: PinT<I, $Mode1>) -> Self {
+                    pin.into_mode()
+                }
+            }
+            $( #[$cfg2] )?
+            impl<I> From<PinT<I, $Mode2>> for PinT<I, $Mode1>
+            where
+                I: PinId,
+            {
+                #[inline]
+                fn from(pin: PinT<I, $Mode2>) -> Self {
+                    pin.into_mode()
+                }
+            }
+        )*
+        impl_core_convert_from!(
+            $(
+                $( #[$cfg2] )?
+                $Mode2,
+            )*
+        );
+    };
+}
+
+impl_core_convert_from!(
+    FloatingDisabled,
+    PullDownDisabled,
+    PullUpDisabled,
+    FloatingInput,
+    PullDownInput,
+    PullUpInput,
+    PushPullOutput,
+    ReadableOutput,
+    AlternateA,
+    AlternateB,
+    AlternateC,
+    AlternateD,
+    AlternateE,
+    AlternateF,
+    AlternateG,
+    #[cfg(any(feature = "samd21", feature = "min-samd51g"))]
+    AlternateH,
+    #[cfg(feature = "min-samd51g")]
+    AlternateI,
+    #[cfg(feature = "min-samd51g")]
+    AlternateJ,
+    #[cfg(feature = "min-samd51g")]
+    AlternateK,
+    #[cfg(feature = "min-samd51g")]
+    AlternateL,
+    #[cfg(feature = "min-samd51g")]
+    AlternateM,
+    #[cfg(feature = "min-samd51g")]
+    AlternateN,
+);
+
+//==============================================================================
+//  Pin IDs
+//==============================================================================
+
+// Because the `Group` trait gives access to the raw registers, hide it in a
+// private module
+mod private {
+    use super::{Sealed, PORT};
+
+    #[cfg(any(feature = "samd11", feature = "samd21"))]
+    use crate::target_device::port::{
+        CTRL, DIR, DIRCLR, DIRSET, DIRTGL, IN, OUT, OUTCLR, OUTSET, OUTTGL, PINCFG0_ as PINCFG,
+        PMUX0_ as PMUX, WRCONFIG,
+    };
+
+    #[cfg(feature = "min-samd51g")]
+    use crate::target_device::port::group::{
+        CTRL, DIR, DIRCLR, DIRSET, DIRTGL, IN, OUT, OUTCLR, OUTSET, OUTTGL, PINCFG, PMUX, WRCONFIG,
+    };
+
+    /// The SAMD11 and SAMD21 PACs don't have the GROUP type.
+    /// Manually re-implement it here
+    pub struct GROUP {
+        pub dir: DIR,
+        pub dirclr: DIRCLR,
+        pub dirset: DIRSET,
+        pub dirtgl: DIRTGL,
+        pub out: OUT,
+        pub outclr: OUTCLR,
+        pub outset: OUTSET,
+        pub outtgl: OUTTGL,
+        pub in_: IN,
+        pub ctrl: CTRL,
+        pub wrconfig: WRCONFIG,
+        _padding1: [u8; 4],
+        pub pmux: [PMUX; 16],
+        pub pincfg: [PINCFG; 32],
+        _padding2: [u8; 32],
+    }
+
+    /// Trait used to implement zero-sized references to the GROUP registers
+    pub trait Group: Sealed {
+        const NUM: usize;
+        #[inline]
+        fn group() -> &'static GROUP {
```

Comment:
I think this is technically `unsafe` if an `&mut` reference to the group exists when this method is called.

Reply:
This function is intended only for consumption inside this module. The structure of this module guarantees that each pin accesses mutually exclusive registers and fields within the `GROUP`. Furthermore, each pin is a singleton, so there is only ever one reference to each set of those registers/fields. As a consequence, there is no chance for memory unsafety.

However, if used improperly, it could create problems. It sounds like I need to both document my assumptions and also mark the function as `unsafe` to communicate that the future authors.

Reply:
Probably a good idea. If I'm understanding this code and Rust's safety rules correctly, just calling this while some other code is holding onto the mutable reference returned by `Pins::mut_port` would technically be undefined behavior, even if they didn't read from or write to that reference. Whether that would cause issues in practice is uncertain, I suppose.

Reply:
I guess the actual write happens atomically even though the bits are being "set" in separate calls, so the interactions with `WRCONFIG` are probably fine. The rest of this is probably fine too, except perhaps for the (very) theoretical multi-threading issue I noted in my other comment.

Reply:
I don't think the same aliasing arguments apply here, because we're using memory-mapped hardware. See my post in the main discussion thread.

Reply:
I think it's technically UB to hold both mutable and shared references to the struct, regardless of what's in it. But since (as I understand it) all of the registers can be mutated via a shared reference, it seems like there's little reason to take an `&mut GROUP` anyway.

Reply:
I ended up deleting the `Pins::mut_port` function anyway, because I don't think it's needed. There's no reason to mutate `PORT`, because each register is represented as a `VolatileCell` with interior mutability. If we find uses for it later, we can always add it back.